samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala

```diff
@@ -1026,6 +1036,34 @@ class SamzaContainer(
     }
   }
 
+  def startDrainMonitor: Unit = {
+    if (drainMonitor != null) {
+      drainMonitor.registerDrainCallback(new DrainCallback {
+        override def onDrain(): Unit = drain()
+      })
+
+      /**
+       * Prior to starting the DrainMonitor, we are doing a one-time check to see if DrainNotification is present
+       * in the metadata store for the current deployment.
+       * This check is to deal with the case where a container might have re-started during Drain.
+       * If yes, we will set the container to drain mode to prevent it from processing any new messages. This will
+       * in-turn guarantee that intermediate Drain control messages from the previous incarnation of the container are
+       * processed and there are no duplicate intermediate control messages for the same deployment.
+       * */
+      if (drainMonitor.shouldDrain()) {
```

Comment:
Let's do it part of the drainMonitor.start(). So start() will do:
1. do a one-time check to see whether it's draining.
2. If yes, then invoke the callback, no need to schedule the timer
3. If no, schedule the timer.

Let's not expose the internal here.